### DISCUSSION TITLE
Problem with ccache after changing branches with larger diff

### DISCUSSION THREAD

**Discussion by AnDevi** [ccache/ccache#1130] [Support]
I'm compiling my app with ccache enabled. Everything works fine until I change my repo's branch to the one with bigger diff.

Flow:

Compiling branch X for the first time (ccache cached everything) - compilation time about 3min
Compiling branch X again (using cache created in previous step) compilation time about 9 sek
Just changing to branch Y for 1 sek
Changing back to branch X
Compiling branch X again (ccache doesn't hit anymore, it caches files again) - compilation time about 3min
If I change to other branch with much smaller diff, ccache works ok.

I have no clue why it doesn't work in scenario above.

Here are options I'm using:
```
(default) absolute_paths_in_stderr = false
(default) base_dir =
(environment) cache_dir = /ccache
(default) compiler =
(default) compiler_check = mtime
(environment) compiler_type = clang
(default) compression = true
(default) compression_level = 0
(default) cpp_extension =
(environment) debug = true
(environment) debug_dir = /ccache/debug
(environment) depend_mode = true
(environment) direct_mode = true
(default) disable = false
(default) extra_files_to_hash =
(environment) file_clone = true
(default) hard_link = false
(default) hash_dir = true
(default) ignore_headers_in_manifest =
(default) ignore_options =
(environment) inode_cache = true
(default) keep_comments_cpp = false
(default) limit_multiple = 0.8
(environment) log_file = /ccache/ccache-log
(default) max_files = 0
(/ccache/config) max_size = 20.0G
(default) namespace =
(default) path =
(default) pch_external_checksum = false
(default) prefix_command =
(default) prefix_command_cpp =
(default) read_only = false
(default) read_only_direct = false
(default) recache = false
(default) reshare = false
(environment) run_second_cpp = true
(default) secondary_storage =
(environment) sloppiness = clang_index_store, include_file_ctime, include_file_mtime, ivfsoverlay, locale, modules, pch_defines, system_headers, time_macros
(default) stats = true
(default) stats_log =
(default) temporary_dir = /ccache/tmp
(default) umask =
```

Anyone has any idea what am I doing wrong ?

PS. It's probably because of that my .hpp files timestamps changes during that branch switching, but isn't ccache resistant for such a timestamp change ? It's is obvious that during compilation without ccache after such a change program will have compile from scratch, but I think that ccache should help for that. Am I wrong ?

**Comment by jrosdahl** [comment]
Hi,

See the [_Cache debugging_](https://ccache.dev/manual/4.6.1.html#_cache_debugging) chapter (especially the last part) in the documentation for tips on how to investigate why you don’t get cache hits.

**Reply by AnDevi** [reply]
Yes, I've already read that. I noticed that during this unlucky long recompilation in `*.ccache-input-text` file  I have additional part `=== PREPROCESSOR MODE ===`, which maybe tells me that timestamps of headers used in my `pch` changed, that why it have rebuild everything from scratch ? 

While only difference I see in `*.ccache-log` is that part: 
```
Retrieved b653hdg1ouam4f380mo271hlfflep66uo from primary storage (/ccache/b/6/53hdg1ouam4f380mo271hlfflep66uoM)
Looking for result key in /ccache/b/6/53hdg1ouam4f380mo271hlfflep66uoM
Did not find result key in manifest
```

which is weird to me, cuz before this changing of branches same key `b653hdg1ouam4f380mo271hlfflep66uo` was found in same cached file `53hdg1ouam4f380mo271hlfflep66uoM`..

**Reply by jrosdahl** [reply]
> Yes, I've already read that.

OK. That would have been good to know. :wink:

>  I noticed that during this unlucky long recompilation in `*.ccache-input-text` file I have additional part `=== PREPROCESSOR MODE ===`

That's not strange by itself, it's just information that shows up on a cache miss (and on a direct mode cache miss).

> why it have rebuild everything from scratch ?

That's impossible to say without having access to the `*.ccache-*` files and potentially the manifest file referenced by the log. Perhaps you can publish them (for the first and second runs of one single compilation unit) and I can have a look? Alternatively, you can send them to me in private mail (joel@rosdahl.net).

> before this changing of branches same key `b653hdg1ouam4f380mo271hlfflep66uo` was found in same cached file `53hdg1ouam4f380mo271hlfflep66uoM`..

I think you have misinterpreted the log messages. Ccache looks for **a** result key in the manifest identified by b653hdg1ouam4f380mo271hlfflep66uo on both cases. What that result key is can be found later on in the log: it's the name of the result file. Look for "Retrieved <result_key> from primary storage (/ccache/x/x/xxx...xxx**R**)". Unless something is very strange, those result keys are different for the first and second run. You can deduce which header files were different for the two runs by interpreting the output of `ccache --inspect /path/to/manifest/file`.

**Reply by AnDevi** [reply]
Hi, I created minimalistic project where you can face this problem:
https://github.com/AnDevi/example_project_with_ccache_hitting_problem

I may say that I found out what is the problem. I'm using precompile headers, and I'm compiling them to `.gch`. When timestamp of some header which is included in that `pch` changes there is a problem, cuz newly generated `.gch` differs from previous one and then all `.cpp` files which use this `pch`  have to be recompiled, cuz `ccache` doesn't find hits for them. 

Workaround solution for me in that case is to use `pch` files as prefix headers and just include them instead of using them as precompile headers.

In this project I posted above I alse described how to force this problem to occur in easy steps.

**Reply by jrosdahl** [reply]
Great that you managed to reproduce the problem with a small project!

What needs to be done to use precompiled headers is documented in the ccache manual's [_Precompiled headers_](https://ccache.dev/manual/4.6.1.html#_precompiled_headers) chapter. I guess the QBS "prefix header" setting results in _"use the compiler option `-include` to include the precompiled header"_.
